JSErrorReport::initBorrowedLinebuf should be called with aligned pointer for char16_t
Sent upstream
https://bugzilla.mozilla.org/attachment.cgi?id=8853441

--- firefox-52.0/js/src/jsexn.cpp.orig	2017-03-31 05:34:08.202506586 +0000
+++ firefox-52.0/js/src/jsexn.cpp	2017-03-31 06:59:30.734728725 +0000
@@ -174,8 +174,10 @@
     /*
      * The mallocSize can not overflow since it represents the sum of the
      * sizes of already allocated objects.
+     * Plus one is to allow skip one byte when the messageSize is odd because
+     * some platforms need to have char16_t* even.
      */
-    size_t mallocSize = sizeof(JSErrorReport) + messageSize + linebufSize + filenameSize;
+    size_t mallocSize = sizeof(JSErrorReport) + messageSize + linebufSize + filenameSize + 1;
     uint8_t* cursor = cx->pod_calloc<uint8_t>(mallocSize);
     if (!cursor)
         return nullptr;
@@ -190,6 +192,8 @@
     }
 
     if (report->linebuf()) {
+        if ((size_t)cursor % 2)
+            cursor++;
         const char16_t* linebufCopy = (const char16_t*)cursor;
         js_memcpy(cursor, report->linebuf(), linebufSize);
         cursor += linebufSize;
